latch_cli/centromere/ctx.py

```diff
@@ -0,0 +1,264 @@
+import re
+import subprocess
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Dict, Optional
+
+import docker
+import paramiko
+from flytekit.core.base_task import PythonTask
+from flytekit.core.context_manager import FlyteEntities
+
+import latch_cli.tinyrequests as tinyrequests
+from latch_cli.centromere.utils import (
+    construct_dkr_client,
+    construct_ssh_client,
+    import_flyte_objects,
+)
+from latch_cli.config.latch import LatchConfig
+from latch_cli.utils import (
+    account_id_from_token,
+    current_workspace,
+    generate_temporary_ssh_credentials,
+    hash_directory,
+    retrieve_or_login,
+)
+
+config = LatchConfig()
+endpoints = config.sdk_endpoints
+
+
+@dataclass
+class Container:
+    dockerfile: Path
+    image_name: str
+
+
+class CentromereCtx:
+    """Manages state for interaction with centromere.
+
+    The context holds values that are relevant throughout the "lifetime" of a
+    registration or remote execution, eg. location of local code and
+    package name, as well as managing docker, ssh clients.
+    """
+
+    dkr_repo: Optional[str] = None
+    dkr_client: docker.APIClient = None
+    ssh_client: paramiko.client.SSHClient = None
+    pkg_root: Path = None  # root
+    ssh_key_path: Path = None
+    disable_auto_version: bool = False
+    image_full = None
+    token = None
+    version = None
+    serialize_dir = None
+    default_container: Container
+    # Used to asscociate alternate containers with tasks
+    container_map: Dict[str, Container] = {}
+
+    latch_register_api_url = endpoints["register-workflow"]
+    latch_image_api_url = endpoints["initiate-image-upload"]
+    latch_provision_url = endpoints["provision-centromere"]
+    latch_get_image_url = endpoints["get-image-from-task"]
+
+    def __init__(
+        self,
+        pkg_root: Path,
+        token: Optional[str] = None,
+        disable_auto_version: bool = False,
+        remote: bool = False,
+    ):
+        try:
+            if token is None:
+                self.token = retrieve_or_login()
+            else:
+                self.token = token
+
+            ws = current_workspace()
+            if ws == "" or ws is None:
+                self.account_id = account_id_from_token(self.token)
+            else:
+                self.account_id = ws
+
+            self.pkg_root = Path(pkg_root).resolve()
+            self.disable_auto_version = disable_auto_version
+            try:
+                version_file = self.pkg_root.joinpath("version")
+                with open(version_file, "r") as vf:
+                    self.version = vf.read().strip()
+                if not self.disable_auto_version:
+                    hash = hash_directory(self.pkg_root)
+                    self.version = self.version + "-" + hash[:6]
```

Comment:
add back check if the version has already been registered (we can make this an api call for consistency across machines)

Reply:
im going to do this as a nucleus call